staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go

```diff
@@ -242,6 +245,15 @@ func (s *Serializer) doEncode(obj runtime.Object, w io.Writer) error {
 		_, err = w.Write(data)
 		return err
 	}
+	if s.options.StreamingCollectionsEncoding {
+		ok, err := streamingEncode(obj, w)
+		if err != nil {
+			return err
```

Comment:
https://github.com/kubernetes/kubernetes/blob/b4f902f0371485505ff4eda39975e67bfa9b0727/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go#L231

does it need to consider adjust for gzip? the whole obj size can trigger gzip but first write `{` is less than defaultGzipSize. it won't enable gzip for that response.

Reply:
> Such change should be done independently to streaming

Agree. Thanks @serathius this change is really great!

Reply:
> Nice catch, independent of streaming this seems like bad decision to depend on internal implementation details in encoder. Such change should be done independently to streaming. We just need to improve gzip to buffer first `defaultGzipThresholdBytes` bytes.

Agree on needing the fix, agree on it being separate, but that would be a prereq to enabling this by default. Can you capture that as a one-line description in the corresponding KEP?

Reply:
Sounds good